pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/RealtimeKinesisIntegrationTest.java

```diff
@@ -0,0 +1,448 @@
+        .setSegmentVersion(getSegmentVersion()).setLoadMode(getLoadMode()).setTaskConfig(getTaskConfig())
+        .setBrokerTenant(getBrokerTenant()).setServerTenant(getServerTenant()).setIngestionConfig(getIngestionConfig())
+        .setLLC(true).setStreamConfigs(createKinesisStreamConfig()).setNullHandlingEnabled(getNullHandlingEnabled())
+        .build();
+  }
+
+  public Map<String, String> createKinesisStreamConfig() {
+    Map<String, String> streamConfigMap = new HashMap<>();
+    String streamType = "kinesis";
+    streamConfigMap.put(StreamConfigProperties.STREAM_TYPE, streamType);
+
+    streamConfigMap
+        .put(StreamConfigProperties.constructStreamProperty(STREAM_TYPE, StreamConfigProperties.STREAM_TOPIC_NAME),
+            STREAM_NAME);
+
+    streamConfigMap.put(
+        StreamConfigProperties.constructStreamProperty(STREAM_TYPE, StreamConfigProperties.STREAM_FETCH_TIMEOUT_MILLIS),
+        "30000");
+    streamConfigMap
+        .put(StreamConfigProperties.constructStreamProperty(STREAM_TYPE, StreamConfigProperties.STREAM_CONSUMER_TYPES),
+            StreamConfig.ConsumerType.LOWLEVEL.toString());
+    streamConfigMap.put(StreamConfigProperties
+            .constructStreamProperty(STREAM_TYPE, StreamConfigProperties.STREAM_CONSUMER_FACTORY_CLASS),
+        KinesisConsumerFactory.class.getName());
+    streamConfigMap
+        .put(StreamConfigProperties.constructStreamProperty(STREAM_TYPE, StreamConfigProperties.STREAM_DECODER_CLASS),
+            "org.apache.pinot.plugin.stream.kafka.KafkaJSONMessageDecoder");
+    streamConfigMap.put(KinesisConfig.REGION, REGION);
+    streamConfigMap.put(KinesisConfig.MAX_RECORDS_TO_FETCH, String.valueOf(MAX_RECORDS_TO_FETCH));
+    streamConfigMap.put(KinesisConfig.SHARD_ITERATOR_TYPE, ShardIteratorType.AT_SEQUENCE_NUMBER.toString());
+    streamConfigMap.put(KinesisConfig.ENDPOINT, LOCALSTACK_KINESIS_ENDPOINT);
+    streamConfigMap.put(KinesisConfig.ACCESS_KEY, getLocalAWSCredentials().resolveCredentials().accessKeyId());
+    streamConfigMap.put(KinesisConfig.SECRET_KEY, getLocalAWSCredentials().resolveCredentials().secretAccessKey());
+    streamConfigMap.put(StreamConfigProperties.SEGMENT_FLUSH_THRESHOLD_ROWS, Integer.toString(5000));
+    streamConfigMap.put(StreamConfigProperties
+        .constructStreamProperty(streamType, StreamConfigProperties.STREAM_CONSUMER_OFFSET_CRITERIA), "smallest");
+    return streamConfigMap;
+  }
+
+  public void startKinesis()
+      throws Exception {
+    try {
+      final LocalstackDockerConfiguration dockerConfig = PROCESSOR.process(this.getClass());
+      localstackDocker.startup(dockerConfig);
+    } catch (LocalstackDockerException e) {
+      StopAllLocalstackDockerCommand stopAllLocalstackDockerCommand = new StopAllLocalstackDockerCommand();
+      stopAllLocalstackDockerCommand.execute();
+
+      final LocalstackDockerConfiguration dockerConfig = PROCESSOR.process(this.getClass());
```

Comment:
why trying to startup again in the catch block?

Reply:
Sometimes in teardown the localstack docker container is not killed. So at the time of running the test again, the localstack startup fails since there is already a container running occupying the same ports.
So here we kill all the running localstack containers and then retry once more. If it fails, then the test is terminated.

We can't use the already running localstack container since it may contain some data already in the kinesis stream.